substrate/primitives/runtime/src/generic/unchecked_extrinsic.rs

```diff
@@ -33,16 +34,199 @@ use codec::{Compact, Decode, Encode, EncodeLike, Error, Input};
 use sp_io::hashing::blake2_256;
+use sp_weights::Weight;
+
+/// Type to represent the version of the [Extension](TransactionExtension) used in this extrinsic.
+pub type ExtensionVersion = u8;
+/// Type to represent the extrinsic format version which defines an [UncheckedExtrinsic].
+pub type ExtrinsicVersion = u8;
 
 /// Current version of the [`UncheckedExtrinsic`] encoded format.
 ///
 /// This version needs to be bumped if the encoded representation changes.
 /// It ensures that if the representation is changed and the format is not known,
 /// the decoding fails.
-const EXTRINSIC_FORMAT_VERSION: u8 = 4;
+pub const EXTRINSIC_FORMAT_VERSION: ExtrinsicVersion = 5;
+/// Current version of the [Extension](TransactionExtension) used in this
+/// [extrinsic](UncheckedExtrinsic).
+///
+/// This version needs to be bumped if there are breaking changes to the extension used in the
+/// [UncheckedExtrinsic] implementation.
+pub const EXTENSION_VERSION: ExtensionVersion = 0;
 
 /// The `SignaturePayload` of `UncheckedExtrinsic`.
-type UncheckedSignaturePayload<Address, Signature, Extra> = (Address, Signature, Extra);
+pub type UncheckedSignaturePayload<Address, Signature, Extension> = (Address, Signature, Extension);
+
+impl<Address: TypeInfo, Signature: TypeInfo, Extension: TypeInfo> SignaturePayload
+	for UncheckedSignaturePayload<Address, Signature, Extension>
+{
+	type SignatureAddress = Address;
+	type Signature = Signature;
+	type SignatureExtra = Extension;
+}
+
+/// A "header" for extrinsics leading up to the call itself. Determines the type of extrinsic and
+/// holds any necessary specialized data.
+#[derive(Eq, PartialEq, Clone)]
+pub enum Preamble<Address, Signature, Extension> {
+	/// An extrinsic without a signature or any extension. This means it's either an inherent or
+	/// an old-school "Unsigned" (we don't use that terminology any more since it's confusable with
+	/// the general transaction which is without a signature but does have an extension).
+	///
+	/// NOTE: In the future, once we remove `ValidateUnsigned`, this will only serve Inherent
+	/// extrinsics and thus can be renamed to `Inherent`.
+	Bare(ExtrinsicVersion),
+	/// An old-school transaction extrinsic which includes a signature of some hard-coded crypto.
+	Signed(Address, Signature, ExtensionVersion, Extension, ExtrinsicVersion),
+	/// A new-school transaction extrinsic which does not include a signature.
+	General(ExtensionVersion, Extension),
+}
+
+const VERSION_MASK: u8 = 0b0011_1111;
+const TYPE_MASK: u8 = 0b1100_0000;
+const BARE_EXTRINSIC: u8 = 0b0000_0000;
+const SIGNED_EXTRINSIC: u8 = 0b1000_0000;
+const GENERAL_EXTRINSIC: u8 = 0b0100_0000;
+
+const LOWEST_SUPPORTED_EXTRINSIC_FORMAT_VERSION: ExtrinsicVersion = 4;
+
+impl<Address, Signature, Extension> Decode for Preamble<Address, Signature, Extension>
+where
+	Address: Decode,
+	Signature: Decode,
+	Extension: Decode,
+{
+	fn decode<I: Input>(input: &mut I) -> Result<Self, Error> {
```

Comment:
Just to bring up; this decode logic is not quite compatible with what's in the original PR command which says:

```
- 0b00000100: Bare (used to be called "Unsigned"): contains Signature or Extra (extension data). After bare transactions are no longer supported, this will strictly identify an Inherents only.
- 0b10000100: Old-school "Signed" Transaction: contains Signature and Extra (extension data).
- 0b01000100: New-school "General" Transaction: contains Extra (extension data), but no Signature.
```

It looks like we actually have:

- 0b00000100 Bare (as above)
- 0b10000100 Old-school "signed" (as above)
- 0b10000101 New-school "signed" transactions with an extension version byte (not mentioned above)
- 0b01000101 New-school "General" transactions (version is 5, not 4 as written above)

Just want to check that it's all good; we should either update the comment or the code :)

Reply:
The PR description was outdated. I changed it now, essentially we preserve backwards compatibility with extrinsic version 4 in terms of decoding, but the current version is now 5 and it adheres to the updated description (as well as [RFC84](https://github.com/polkadot-fellows/RFCs/pull/84)) and the decoding logic [here](https://github.com/paritytech/polkadot-sdk/blob/ff0f75664fe46d3945af89cca75399e4937c8908/substrate/primitives/runtime/src/generic/unchecked_extrinsic.rs#L99).

Reply:
That makes sense, thanks!

it looks like the description isn't quite right yet as it says:

"0b10000101: Old-school "Signed" Transaction: contains Signature and Extra (extension data)."

Where I think we want:

"
0b10000100: Old-school "Signed" Transaction: contains Signature and Extra (extension data).
0b10000101: New-school "signed" Transaction: containing an extension version byte in addition to the above.
"

Reply:
I updated the description.

Just to clear up some naming issues, old-school signed transactions exist in both extrinsic verison 4 and 5, but in version 5 we also introduced the extension version byte, described in [RFC99](https://github.com/polkadot-fellows/RFCs/blob/main/text/0099-transaction-extension-version.md).

New-school (general) transactions *could* be "signed" with the appropriate extension, but not necessarily. In theory, we should aim to deprecate signed transaction altogether in favor of general transactions. This is just a transition period.

Reply:
Thank you, that makes sense! I guess this implies that there will also soon be a new TransactionExtension created to handle signing transactions sing the existing approach, which would allow clients to move towards using that rather than the hardcoded signature field?

I sortof wonder: if the plan is to deprecate the "new" signed transactions then is there a reason to have them at all rather than just migrating people from "old" signed transactions to General transactions once ready? Not having a "new signed transaction" at all would mean one less type of thing for client libraries to have to think about being able to decode :)

Reply:
This PR changes some fundamental building blocks of FRAME. First off, there needs to be a transition period simply because people consuming substrate wrote `SignedExtensions` on the old interface and would need a transition period to adapt them to the new format. Meanwhile, they will use `AsTransactionExtension`. It's not only about the signed transactions, it concerns unsigned transactions as well which will be deprecated in phase 2. Inherents will be the only `Bare` transactions, while other currently unsigned transactions will need to go through the "General" pipeline (not signed). Users need time to adapt their existing (signed) extensions to handle the cases where an origin is something else other than `RawOrigin::Signed`. Also, the deprecation of signed transactions will make the signing scheme explicit (through a new extension) rather than implicit (through the `Checkable` impl of `UncheckedExtrinsic`), and that shouldn't happen before all existing extensions are migrated AND we provide a solution for the unsigned use case in phase 2. More details about this can be found in the original Extrinisc Horizon issue, linked [here](https://github.com/paritytech/polkadot-sdk/issues/2415).

Reply:
Thanks for the response!

I understand that in the future we'll have juyst Bare and General Transactions, but the original issue etc doesn't really explain (to me at least) why we need a "new" Signed Transaction as a stepping stone between old ones and General Transactions (since old Signed Transactions will continue to be supported too). So I really just wanted to understand if that "new" Signed Transaction is necessary because, if not, it'd be simpler for client tooling to not have to think about it as a possible variation to encode/decode in addition to old Signed Transactions and new General Transactions.

Nevertheless, if it is necessary to have this "new" Signed Transaction (which will then be deprecated as we move to General Transactions) then I understand; just trying to keep life ismpler for the client libraries that have to encode and decode these things :)